posts/2018-06-15-typedapi.md

```diff
@@ -0,0 +1,329 @@
+---
+layout: post
+title: Typedapi or how to derive your clients and servers from types
+
+meta:
+  nav: blog
+  author: pheymann
+  pygments: true
+---
+
+In this blog post, I will show you how to leverage Scala's type system to derive an HTTP client function from a single type. This will also be the story of how I started to work on [Typedapi](https://github.com/pheymann/typedapi) which is basically the attempt to bring Haskell's [Servant](https://github.com/haskell-servant/servant) to Scala.
+
+## Servant in a nutshell and how it began
+For everyone not knowing Servant, it is a library which lets you define your web apis as types and derives the client and server functions from it. When I saw it for the first time while working on a ped project I immediately loved the idea. Creating web server and clients this way reduces your code to a mere type, you get extra type safety and you can use the api types as contracts between your server and its clients. 
+
+I couldn't find any viable alternative in Scala at the time and decided to build it on my own. But I just wanted to start with a single feature to not overwhelm myself and abandon the project after a short time. Therefore, I set out to make Scala able to derive a client function from a single api type, as we will do in this post.
+
+## Derive a client function from a type. How hard can it be?
+Let's start with an example we will use later on to ease understanding. Consider the following api:
+
+```
+GET /users/:name?minAge=:age -> List[User]
+```
+
+It only consists of a single endpoint which returns a list of `Users`:
+
+```scala
+final case class User(name: String, age: Int)
+```
+
+with a given `name: String`. Furthermore, you filter the resulting users by their `age: Int`. Our big goal is to end up with a function which is derived from a type-level representation of our endpoint:
+
+```scala
+(name: String, minAge: Int) => F[List[User]]
+```
+
+### Represent the api as a type
+Question: how do you represent the above api as a type in Scala? I think the best way is to break it apart and try to find type-level representations for each element. After that, we "just" merge them together.
+
+When we take a closer look at our endpoint we see that it consists of:
+  * path elements identifying an endpoint: `/users`
+  * segments within the path: `:name`
```

Comment:
For segments and queries, it might be worth emphasizing that we need them to represent a parameter which has a name *and* a type